File: 09Function/closure/input_functions.rs
//这个应该在C#中用的delegate的方式，对比一下，C#的delegate支持闭包吗？试试。

// 定义一个函数，可以接受一个由 `Fn` 限定的泛型 `F` 参数并调用它。
fn call_me<F: Fn()>(f: F) {
    f()
}

// 定义一个满足 `Fn` 约束的封装函数（wrapper function）。
fn function() {
    println!("I'm a function!");
}

fn fib_root<F:Fn(i32,i32,i32)->i32>(f:F,n:i32){
	println!("call a recursive function! {} ", f(n,1,1));
}

fn fib(n:i32,pre:i32,ppre:i32)->i32 {
	if n == 1 || n == 2 {
		pre
	}else {
		fib( n-1, pre+ppre, pre)
	}
}

fn main() {
    // 定义一个满足 `Fn` 约束的闭包。
    let closure = || println!("I'm a closure!");
    
    call_me(closure);
    call_me(function);
    fib_root(fib, 5);
}

File: 09Function/closure/input_functions.cs
using System;
using System.Diagnostics;
public class Program
{
    // 定义一个函数，可以接受一个由 `Fn` 限定的泛型 `F` 参数并调用它。
    static void call_me(Action f) {
        f();
    }

    // 定义一个满足 `Fn` 约束的封装函数（wrapper function）。
    static void function() {
        Console.WriteLine("I'm a function!");
    }

    static void fib_root(Func<int,int,int,int>f,int n ) {
        Console.WriteLine("call a recursive function! {0} ", f(n,1,1));
    }

    static int fib(int n,int pre , int ppre) => 
        n switch {
            (1 or 2)=>pre,
            _ => fib(n-1, pre+ppre, pre),
        };
    public static void Main(string[] args)
    {
		Console.WriteLine("This is C# Language");
        // 定义一个满足 `Fn` 约束的闭包。
        Action closure = ()=>Console.WriteLine("I'm a closure!");
        
        call_me(closure);
        call_me(function);
        Stopwatch stopwatch = Stopwatch.StartNew();
        fib_root(fib, 45);
        stopwatch.Stop();
        // 如果你需要更精细的时间单位，可以这样做：
        Console.WriteLine("Nanoseconds elapsed: {0}", (double)stopwatch.ElapsedTicks/Stopwatch.Frequency * 1_000_000_000);
    }
}
/*
I'm a closure!
I'm a function!
call a recursive function! 1134903170
Elapsed milliseconds: 335500 //C# 这个数值为308600，尾递归只防止overflow 但是执行时间并没有少太多啊。
*/